internal/sql/pg/repl_changeset.go

```diff
@@ -1,29 +1,253 @@
 
-	writer io.Writer
+	// the io.Writer is multipurpose:
+	//  - for capturing and transmitting the changeset in long running network migrations
+	//  - updating table statistics required for query cost
+	// writer io.Writer
+
+	csChan chan<- any // *Relation / *ChangesetEntry
 }
 
 var (
-	changesetInsertByte   = byte(0x01)
-	changesetUpdateByte   = byte(0x02)
-	changesetDeleteByte   = byte(0x03)
-	changesetMetadataByte = byte(0x04)
+	RelationType       = byte(0x01)
+	ChangesetEntryType = byte(0x02)
+	BlockSpendsType    = byte(0x03)
 )
 
+type ChangesetEntry struct {
+	// RelationIdx is the index in the full relation list for the changeset that
+	// precedes the tuple change entries.
+	RelationIdx uint32
+
+	// userSchema: set to true if the schema is user-defined
+	UserSchema bool
+
+	OldTuple []*TupleColumn // empty for insert
+	NewTuple []*TupleColumn // empty for delete
+	// both old and new are set for update
+}
+
+func (ce *ChangesetEntry) Kind() string {
+	if len(ce.NewTuple) == 0 {
+		return "delete"
+	}
+	if len(ce.OldTuple) == 0 {
+		return "insert"
+	}
+	return "update"
+}
+
+func (ce *ChangesetEntry) String() string {
+	return fmt.Sprintf("Change type %s, rel ID %d, %d old tuples, %d new tuples",
+		ce.Kind(), ce.RelationIdx, len(ce.OldTuple), len(ce.NewTuple))
+}
+
+func (ce *ChangesetEntry) Serialize() ([]byte, error) {
+	buf := new(bytes.Buffer)
+	buf.WriteByte(ChangesetEntryType)
+
+	bts, err := serialize.Encode(ce)
+	if err != nil {
+		return nil, err
+	}
+
+	size := uint32(len(bts))
+	err = binary.Write(buf, binary.LittleEndian, size)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write size: %w", err)
+	}
+
+	_, err = buf.Write(bts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write tuple data: %w", err)
+	}
+
+	return buf.Bytes(), nil
+}
+
+func (ce *ChangesetEntry) Deserialize(data []byte) error {
+	return serialize.Decode(data, ce)
+}
+
+func (ce *ChangesetEntry) ApplyChangesetEntry(ctx context.Context, tx sql.DB, relation *Relation) error {
+	switch ce.Kind() {
+	case "insert":
+		return ce.applyInserts(ctx, tx, relation)
+	case "delete":
+		return ce.applyDeletes(ctx, tx, relation)
+	default:
+		return ce.applyUpdates(ctx, tx, relation)
+	}
+}
+
+// DecodeTuple decodes serialized tuple column values into their native types.
+// Any value may be nil, depending on the ValueType.
+func (c *ChangesetEntry) DecodeTuples(relation *Relation) (oldValues, newValues []any, err error) {
+	if oldValues, err = decodeTuple(c.OldTuple, relation); err != nil {
+		return nil, nil, err
+	}
+
+	if newValues, err = decodeTuple(c.NewTuple, relation); err != nil {
+		return nil, nil, err
+	}
+
+	return oldValues, newValues, nil
+}
+
+func decodeTuple(cols []*TupleColumn, relation *Relation) ([]any, error) {
+	if cols == nil {
+		return nil, nil
+	}
+
+	values := make([]any, len(cols))
+	for i, col := range cols {
+		switch col.ValueType {
+		case NullValue, ToastValue:
+			values[i] = nil
```

Comment:
This is OK, but I omitted it because it's superfluous.  `values := make([]any, len(cols))` puts the zero value in each element on construction. (nil)

Reply:
BTW, have we tested with TOASTed values.  By default that's records over 4096, but we should also try with really large values, particularly for the "old" tups in update or delete.

Reply:
I hope that with replica identity full we get the full values, but I don't know.